oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/geometry/OSHDBGeometryBuilder.java

```diff
@@ -368,47 +449,84 @@ private static Polygon constructMultipolygonPart(
+   *                    polygonal geometry for the respective entity (based on its tags)
+   * @param clipPoly a polygon to clip the resulting geometry to
+   * @param <P> either {@link Polygon} or {@link org.locationtech.jts.geom.MultiPolygon}
+   * @return a JTS geometry object (simple features compatible, i.e. a Point, LineString, Polygon
+   *         or MultiPolygon)
+   */
+  public static <P extends Geometry & Polygonal> Geometry getGeometryClipped(
+      OSMEntity entity, OSHDBTimestamp timestamp, TagInterpreter areaDecider, P clipPoly
+  ) {
     Geometry geom = OSHDBGeometryBuilder.getGeometry(entity, timestamp, areaDecider);
     return Geo.clip(geom, clipPoly);
   }
-  
+
   /**
-   * Converts a OSHDBBoundingBox to a rectangular polygon.
-   *
-   * <p>
-   * Will return a polygon with exactly 4 vertices even for point or line-like BoundingBox.
-   * Nevertheless, for degenerate bounding boxes (width and/or height of 0) the result might
-   * not pass the {@link Geometry#isRectangle() Geometry.isRectangle} test.
-   * </p>
+   * Converts a JTS bounding box ("envelope") to an OSHDBBoundingBox object.
    *
-   * @param bbox The BoundingBox the polygon should be created for.
-   * @return a rectangular Polygon
+   * @param envelope the bounding box object to convert
+   * @return the same bounding box as an OSHDBBoundingBox object
    */
-  public static Polygon getGeometry(OSHDBBoundingBox bbox) {
-    assert bbox != null : "a bounding box is not allowed to be null";
-    
-    GeometryFactory gf = new GeometryFactory();
-        
-    Coordinate sw = new Coordinate(bbox.getMinLon(), bbox.getMinLat());
-    Coordinate se = new Coordinate(bbox.getMaxLon(), bbox.getMinLat());
-    Coordinate nw = new Coordinate(bbox.getMaxLon(), bbox.getMaxLat());
-    Coordinate ne = new Coordinate(bbox.getMinLon(), bbox.getMaxLat());
-
-    Coordinate[] cordAr = {sw, se, nw, ne, sw};
-
-    return gf.createPolygon(cordAr);
-  }
-  
   public static OSHDBBoundingBox boundingBoxOf(Envelope envelope) {
-    return new OSHDBBoundingBox(envelope.getMinX(), envelope.getMinY(), envelope.getMaxX(), envelope.getMaxY());
+    return new OSHDBBoundingBox(
+        envelope.getMinX(),
+        envelope.getMinY(),
+        envelope.getMaxX(),
+        envelope.getMaxY()
+    );
   }
 
+  private static class Segment {
+    long id1;
+    long id2;
+
+    Segment(long id1, long id2) {
+      this.id1 = id1;
+      this.id2 = id2;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+      if (other instanceof Segment) {
+        Segment otherSegment = (Segment) other;
+        return otherSegment.id1 == this.id1 && otherSegment.id2 == this.id2
+            || otherSegment.id1 == this.id2 && otherSegment.id2 == this.id1;
+      } else {
+        return super.equals(other);
+      }
+    }
+
+    @Override
+    public int hashCode() {
+      return ((int) this.id1) + ((int) this.id2);
```

Comment:
Are the here used ids always hashes? Is the addition unique enough?

Reply:
The ids are just the osm ids of the nodes of the respective way segment. I think I just wanted something quick satisfies equality regardless of the order of the ids. Could this result in issues down the line, e.g. when used in a HashMap?